### ISSUE TITLE
Train and lock feature profile models

### ISSUE THREAD

**Issue by antonstrover** [antonstrover/Avalanche#235]
## Purpose

Train and freeze one independent model for each approved principal profile.

Preserve a negative result when no candidate passes the unchanged gate.

## Audit findings addressed

- AUDIT-02 §8.1 item 2, principal model fails its gate: primary owner.
- AUDIT-02 §8.1 item 3, learned configuration cannot load: primary owner.
- AUDIT-02 §8.2 item 6, runtime bypasses locks: supporting work.
- AUDIT-02 §4.5, failed models and synthetic acceptance: primary owner.
- AUDIT-02 §6.4, feature profiles are not controlled ablations: primary owner.
- AUDIT-02 §3 item 12, reproducible principal model: primary owner.

## Decisions implemented

- Decision 9: use ordered attempts and freeze the first pass.
- Decision 10: retrain and calibrate each profile.
- Decision 21: lock every artifact and load offline.
- No project-level CHOICES.md file exists.

## Verified current state

The recorded perceptron and GRU both fail sleeper recall.

The training function raises before preserving failed weight files.

The current final evaluator reuses one model with zeroed blocks.

configs/monitors/learned.yaml points to a missing incompatible model.

Runtime loading accepts direct paths and threshold overrides.

Model locks omit several required identities.

No passing principal artifact exists in a clean checkout.

The final family has already influenced prior candidate evaluation.

## Required end state

Freeze this candidate order:

| Order | Name | Architecture | Seed | Epochs | Batch | Learning rate |
|---:|---|---|---:|---:|---:|---:|
| 1 | mlp-64x32-v3 | MLP 64,32 | 20260901 | 80 | 256 | 0.001 |
| 2 | mlp-128x64-v3 | MLP 128,64 | 20260902 | 80 | 256 | 0.0005 |
| 3 | gru32-window8-v3 | GRU 32, window 8 | 20260903 | 60 | 256 | 0.001 |

Freeze candidate work at 2026-09-30T23:59:59Z.

Consume the immutable candidate registry committed by commit 012.

Do not add a candidate after the first validation result.

Run candidates separately for each of the five principal profiles.

Name each attempt `<profile>--<candidate>`.

Use that identity in lock paths and release asset paths.

Use identical development manifests, roots, cells, labels, and policies.

Train only from each profile's actual columns.

Calibrate each candidate separately.

Write a complete immutable lock for every attempted candidate.

Stop each profile at its first passing candidate.

Keep the sleeper recall gate at 0.800.

Keep the episode false-alarm budget at 0.05.

Rank failures by the greater minimum gate margin.

Define margins as sleeper recall minus 0.800 and 0.05 minus false-alarm rate.

Break failure ties by lower Brier score.

Break later ties by lower calibration error.

Define calibration error as ten-bin expected calibration error.

Use equal-width bins [0.0, 0.1) through [0.9, 1.0].

For each nonempty bin, use its mean score and positive-label rate.

Weight each absolute difference by the bin count divided by all rows.

Omit empty bins from the sum.

Quantize every ranking metric to 12 decimal places with ROUND_HALF_EVEN.

Treat equal quantized values as ties.

Break final ties by candidate order.

For every profile, freeze its first pass or deterministic best failure.

Write one immutable content-addressed selection manifest for every profile.

Store it at `artifacts/monitor/selections/<profile>/<selection_sha256>.json`.

Keep every attempt lock unchanged after creation.

If principal-full has no passing candidate, freeze its best declared failure.

Select that lock with role negative_core_baseline.

Record that role only in the principal-full selection manifest.

Permit only that named failed lock in the final negative learned role.

Make no learned-monitor success claim in that case.

Only principal-full governs the core learned-success claim.

A failed nonprincipal profile can enter only its matching ablation cell.

Label every such cell as a failed-profile negative result.

Keep every other attempt ineligible for formal evaluation.

Never use sealed holdout outcomes for selection.

Reference an exact verified lock file from formal monitor configuration.

Also reference its exact verified selection manifest.

Do not reference a mutable registry entry from formal configuration.

Load model, calibration, threshold, budget, and schema only from that lock.

Publish artifact bytes through immutable release assets.

This repository forbids committed model files.

Therefore, every selected model uses the release-asset path from Decision 21.

Prepare the offline cache explicitly before formal runs.

## Files and symbols affected

- protocols/development/model-candidates-v3.json — consume the registry proposed by commit 012.
- src/avalanche/monitors/perceptron.py — support declared MLP configurations.
- src/avalanche/monitors/training.py — preserve every attempt and selection.
- src/avalanche/monitors/learned.py — load only selected verified locks.
- src/avalanche/monitors/factory.py — consume the verified loader from commit 001.
- src/avalanche/config/models.py — consume the lock_reference type created by commit 001.
- configs/monitors/learned.yaml — populate one exact lock_reference.
- scripts/train_monitor.py — run the frozen candidate registry.
- scripts/compare_monitor_models.py — train independent profile candidates.
- artifacts/monitor/registry-v2.json — extend the proposed registry created by commit 001.
- artifacts/monitor/locks/*.json — create selected and failed lock metadata.
- artifacts/monitor/selections/*.json — create five proposed selection manifests.
- tests/unit/test_training_gate.py — add order and negative-selection tests.
- tests/unit/test_perceptron_classifier.py — remove contaminated random split.
- tests/integration/test_learned_monitor.py — load every profile lock.
- tests/regression/test_gru_ablation.py — retain failed historical evidence only.
- docs/monitor-hardening/README.md — record the new result after verification.

Model and calibration bytes remain release assets.

Do not commit model files.

## Implementation steps

1. Verify the precommitted ordered candidate registry before training.
2. Digest the registry and freeze timestamp.
3. Verify the development manifest and feature registry.
4. Generate one dataset version for the frozen boundary.
5. Train candidate one for principal-full.
6. Calibrate it through the episode-level procedure.
7. Write its artifacts and lock regardless of gate result.
8. Stop after a pass.
9. Otherwise continue through the declared order.
10. Repeat the exact process for every principal profile.
11. Never reuse weights across profiles.
12. Never zero a missing feature.
13. Record every gate component and margin.
14. Quantize every ranking metric through the frozen decimal rule.
15. Select each profile failure only through the frozen ranking.
16. Write one immutable selection manifest for each profile.
17. Record roles only in selection manifests.
18. Mark a selected failed principal-full model explicitly negative.
19. Label failed ablation-profile selections separately.
20. Publish model and calibration bytes as immutable assets.
21. Add their complete SHA-256 values to registry locks.
22. Fetch and verify them into the offline cache.
23. Point learned configuration at the exact lock_reference.
24. Populate threshold, temperature, and schema only from the selected lock.
25. Make formal loading reject every unselected failed lock.
26. Make final reports propagate each profile's status.
27. Preserve old failed baselines from commit 001.
28. Close candidate selection before any sealed access.

### Pre-change failing tests

- tests/unit/test_training_gate.py — proposed test_every_attempt_writes_a_lock.
- tests/unit/test_training_gate.py — proposed test_first_pass_stops_candidate_order.
- tests/unit/test_training_gate.py — proposed test_failure_ranking_is_frozen.
- tests/unit/test_training_gate.py — proposed expected_calibration_error_and_tie_table.
- tests/unit/test_training_gate.py — proposed profile_candidate_attempt_identity_table.
- tests/unit/test_training_gate.py — proposed immutable_selection_manifest_role_table.
- tests/unit/test_training_gate.py — proposed test_only_selected_failure_can_fill_negative_role.
- tests/unit/test_training_gate.py — proposed test_every_profile_selects_pass_or_best_failure.
- tests/unit/test_training_gate.py — proposed test_failed_profile_is_limited_to_matching_ablation.
- tests/integration/test_learned_monitor.py — proposed test_config_references_exact_lock_file.
- tests/unit/test_training_gate.py — proposed test_candidate_after_freeze_is_rejected.
- tests/integration/test_learned_monitor.py — proposed test_each_profile_loads_distinct_model_bytes.
- tests/integration/test_learned_monitor.py — preserve commit 001 runtime override rejection.
- tests/unit/test_perceptron_classifier.py — proposed test_training_uses_manifest_split.

Use small deterministic fixture datasets for order tests.

Do not use sealed families.

### Regression tests

- Preserve the two historical failed results.
- Preserve the 0.800 sleeper gate.
- Preserve the 0.05 episode budget.
- Preserve independent external random streams.
- Preserve whole-run windows.
- Preserve exact artifact verification.

Run:

- uv run pytest tests/unit/test_training_gate.py
- uv run pytest tests/unit/test_perceptron_classifier.py
- uv run pytest tests/integration/test_learned_monitor.py
- uv run pytest tests/regression/test_gru_ablation.py
- uv run pytest tests/unit/test_model_artifacts.py
- uv run ruff check .
- uv run ruff format --check .

The full train_monitor candidate command is proposed by this commit.

Run it only after the protocol and data digests freeze.

## Acceptance criteria

- [ ] The candidate registry has three immutable ordered entries.
- [ ] Candidate selection closes at the exact UTC timestamp.
- [ ] Every attempted candidate has a complete lock.
- [ ] Every attempt identity contains its profile and candidate.
- [ ] Each profile trains from its own feature columns.
- [ ] No model weights are reused across profiles.
- [ ] The first passing candidate freezes immediately.
- [ ] Failure ranking follows the declared margins and ties.
- [ ] Only one named failed lock can serve as the negative core baseline.
- [ ] Every profile selects its first pass or deterministic best failure.
- [ ] Five immutable selection manifests authorize those choices.
- [ ] Failed profile models appear only in matching labelled ablations.
- [ ] A negative selection produces no learned-success claim.
- [ ] Formal configuration references a verified lock.
- [ ] Formal configuration verifies its matching selection manifest.
- [ ] Runtime overrides fail.
- [ ] Formal evaluation works offline from verified cached bytes.
- [ ] No sealed outcome enters selection.

## Dependencies

- 001-preserve-failed-monitor-baselines.md supplies the registry and lock schema.
- 012-freeze-development-and-holdout-manifests.md supplies data and the candidate registry.
- 013-enforce-feature-and-dataset-provenance.md supplies five profile schemas.
- 014-calibrate-episode-level-false-alarms.md supplies threshold selection.

Commit 016 consumes every selected profile lock.

Commit 024 verifies research readiness independently.

## Risks and rollback considerations

No candidate may pass.

The locked negative policy preserves a valid negative dissertation result.

Training can finish after the declared cutoff.

Start only jobs that can complete and publish before the cutoff.

Different hardware can change training bytes.

Use one pinned CPU configuration and one Torch thread.

Release publication can fail after training.

Do not mark selection complete before bytes and locks verify.

Rollback must retain every attempt and cannot rewrite the registry.

## Documentation and evidence updates

Verify the candidate registry published by commit 012 before training.

Publish every result, margin, lock, and creation command.

State clearly when the selected principal model failed.

Do not rewrite historical failure evidence.

Do not report holdout performance in model selection records.

### Included

Include frozen candidate order, independent training, selection, locks, and configuration.

### Excluded

Exclude sealed holdout execution.

Exclude post-hoc occlusion.

Exclude adaptive controller work.

Exclude generic loader, direct-path, and override removal owned by commit 001.

### Repository state after commit

Every principal profile has one immutable selected model or declared negative baseline.
